text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,203 @@
+  the current timestamp by default
+- Delete/BatchDelete - Deletes all versions of a specified key
+  (`key:cur_ts -> delete`)
+- DeleteRange - Deletes all versions of key in a range using rocksdb DeleteRange
+
+The above fetch operations(get/scan) can use `rocksdb prefix_extractor` to
+improve the performance of iterator seek. `key: ts-> delete` indicates that
+ all data before the `ts` has been deleted.
+
+### Version recycling
+
+The version recycling would recycling the redundant versions more than
+`MaxVerNum`, e.g. one key has five versions [v1 v2 v3 v4 v5], and the
+`MaxVerNum` is set to be 3, then [v1 v2] would be recycled.
+
+The design of the version recycling will affect cluster performance and the
+functional designs of incremental backup and CDC. There are two major
+approaches:
+
+- using multi-version Garbage Collection (GC) similar to TxnKV, and
+  periodically starting the gc worker in the background to delete redundant
+  versions of data;
+- using compaction filter via rocksdb compaction to delete redundant versions.
+  The default maximum number of versions is 1, which can default to the
+  compaction filter recycling version , so recycling efficiency is moderate.
+
+The GC method consumes CPU or storage resources, but the version recycling is
+accurate and efficient; the compaction filter method does not take up CPU or
+storage resources. However, compaction only has a local applicable scope, so
+multi-version recycling efficiency is low.
+
+This proposal adopts GC method mainly, and could reuse the logic of TxnKV GC.
+
+### Data increment
+
+The `ScanNumWithSV` rpc interface is used to get all data increments of the
+entire cluster or within a specified range [start_ts, cur_ts). When this
+interface is called, the snapshot of the entire cluster or the current cur_ts
+is obtained before scanning.
+
+VerKV integrated to tidb/br and tidb/ticdc can better facilitate the
+implementation of full/incremental backup and CDC services.
+
+### Data replication
+
+For data replication, there are two main situations:
+
+- A -> B: asynchronous one-way replication, Cluster B accesses the CDC
+  service of cluster A to pull or push incremental data of each period
+- A <-> B: asynchronous two-way replication, which is equivalent to two
+  one-way replication processes. Write conflicts can be resolved according to ts
+
+### Others
+
+Rocksdb User Timestamp（WIP）can also be considered to improve the engine
+efficiency.
+
+## Drawbacks
+
+As compared with RawKV, this VerKV adds version information, so it would
+ decrease the performance of data fetch interface and the version recycling
+would add pressure to system.
+
+And for VerKV is going to replace RawKV, it could not be updated directly
+ from RawKV to VerKV, and have to migrated date to a new cluster with VerKV
+using some method.
+
+Meanwhile, add a new API would add maintenance burden during a period of time
+ before deprecating RawKV in the future.
```

Comment:
```suggestion
Compared with RawKV, this VerKV adds version information, so it would
 decrease the performance of data fetch interface and the version recycling
would add pressure to the system.

Migration from RawKV to VerKV cannot be one directly.  The data has to be  
migrated to a new cluster with VerKV using some extra methods.

Meanwhile, introducing a new API would add maintenance burden for a period of time
before RawKV is deprecated in the future.
```